File: sol/solution/0001-0100/0017. Letter Combinations of a Phone Number/Solution.ts
function letterCombinations(digits: string): string[] {
    if (!digits) return [];
    const map: { [key: string]: string } = {
        "2": "abc",
        "3": "def",
        "4": "ghi",
        "5": "jkl",
        "6": "mno",
        "7": "pqrs",
        "8": "tuv",
        "9": "wxyz"
    };
    const result: string[] = [];
    const backtrack = (comb: string, nextDigits: string) => {
        if (nextDigits.length === 0) {
            result.push(comb);
            return;
        }
        for (const letter of map[nextDigits[0]]) {
            backtrack(comb + letter, nextDigits.slice(1));
        }
    };
    backtrack("", digits);
    return result;
}

File: sol/solution/0001-0100/0017. Letter Combinations of a Phone Number/Solution.dart
class Solution {
  List<String> letterCombinations(String digits) {
    if (digits.isEmpty) return [];
    Map<String, String> phone = {
      '2': 'abc',
      '3': 'def',
      '4': 'ghi',
      '5': 'jkl',
      '6': 'mno',
      '7': 'pqrs',
      '8': 'tuv',
      '9': 'wxyz',
    };
    List<String> result = [];
    void backtrack(String combination, int next) {
      if (next == digits.length) {
        result.add(combination);
        return;
      }
      String letters = phone[digits[next]] ?? '';
      for (int i = 0; i < letters.length; i++) {
        backtrack(combination + letters[i], next + 1);
      }
    }
    backtrack('', 0);
    return result;
  }
}